# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
(lambda num : num + 2)(4)

6

In [3]:
def addTwo(num):
    return num + 2

#### Saving to a Variable

In [2]:
addTwoLambda = lambda num : num + 2

addTwoLambda(19)

21

#### Multiple Inputs

In [4]:
(lambda num1, num2, num3: num1 + num2 + num3)(6, 4, 10)

20

#### Passing a Lambda into a Function

In [8]:
def outerFunction():
    outer_num = 4
    def multiply(num):
        return num * outer_num
    return multiply

print(outerFunction()(4))

my_multiply = outerFunction()

my_multiply(4)

def returnLambda(num1, num2):
    return lambda num3, num4 : num1 + num2 + num3 + num4

addToFive = returnLambda(2, 3)
print(addToFive(10,20))

print(returnLambda(2,3)(10,20))

16
35
35


In [16]:
def how_to_multiply(num1):
    return lambda num2 : num1 * num2

double = how_to_multiply(2)

triple = how_to_multiply(3)

triple(10)

30

#### Returning a Lambda from a Function

In [11]:
def math_operations(f, num):
    return f(num)

math_operations(lambda x : x ** 2, 10)

math_operations(lambda num : num // 10, 1000)

multiply_by_ten = lambda num: num * 10

math_operations(multiply_by_ten, 5)

50

#### If Statements within Lambdas

In [13]:
# if requires else

greater_than_10_add_2 = lambda num : num + 2 if num > 10 else num

greater_than_10_add_2(11)

13

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

Cube if even
Square if odd

In [14]:
f = lambda x : x ** 3 if x % 2 == 0 else x ** 2
f(2)
f(3)
f(-3)

9

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [13]:
# .map(<function>, <iterable>)
# function should return a value

def square_number(num):
    return num**2

number_list = [1, 2, 3, 4, 5, 6]

list(map(square_number, number_list))

[1, 4, 9, 16, 25, 36]

#### Using Lambda's with Map

In [6]:
for num in map(lambda num : num ** 3, number_list):
    print(num, end=' ')
    
[num for num in map(lambda num : num ** 3, number_list)]

1 8 27 64 125 216 

[1, 8, 27, 64, 125, 216]

In [8]:
attackers = [5, 6, 2, 8]
defenders = [4, 7, 1, 7]

list(map(lambda attacker, defender : attacker > defender, attackers, defenders))

[True, False, True, True]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [9]:
number_list = [1, 2, 3, 4, 5, 6]

[num for num in map(lambda num: 2*num-1, number_list)]

[1, 3, 5, 7, 9, 11]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [12]:
# .filter(<function>, <iterable>) returns filter object
# function should return True or False

def greater_than_nine(num):
    return num >= 10

num_list = [10, 1, 100, 5, 11]

[x for x in filter(greater_than_nine, num_list)]

[10, 100, 11]

#### Using Lambda's with Filter()

In [17]:
students = ['Hamed', 'Andy', 'William', 'Christopher', 'Will']
[name for name in filter((lambda student : student[0].lower() == "w"), students)]

['William', 'Will']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [20]:
from statistics import mean

random_numbers = [2, 7, 4.2, 1.6, 9, 4.4, 4.7]
print("Mean ", mean(random_numbers))

#[num for num in filter(lambda num : num < mean(random_numbers), random_numbers)]

list(filter(lambda num : num < mean(random_numbers), random_numbers))

Mean  4.7


[2, 4.2, 1.6, 4.4]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [21]:
from functools import reduce

#.reduce(<function>, <iterable>)
# grab first two elements and combine to one

def add_two_numbers(num1, num2):
    return num1 + num2

reduce(add_two_numbers, [1, 2, 3, 4, 5])

15

#### Using Lambda's with Reduce()

In [22]:
reduce(lambda num1, num2 : num1 * num2, [1, 2, 3, 4, 5])

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to subtract the numbers in the list below together with a lambda function.</p>

In [24]:
reduce(lambda num1, num2: num1 - num2, [10, 9, 8, 7, 6])

-20

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [26]:
def find_factorial(num):
    print()
    if num == 1:
        print(f'find_factorial({num}) = 1')
        return num
    print(f'find_factorial({num}) = {num} + find_factorial({num - 1})')
    return num * find_factorial(num-1)

find_factorial(5)


find_factorial(5) = 5 + find_factorial(4)

find_factorial(4) = 4 + find_factorial(3)

find_factorial(3) = 3 + find_factorial(2)

find_factorial(2) = 2 + find_factorial(1)

find_factorial(1) = 1


120

In [ ]:
def find_factorial(num):
    if num == 1:
        return num
    return num * find_factorial(num-1)

find_factorial(5)

In [28]:
def add_num(num):
    if num == 1:
        return 1
    return num + add_num(num - 1)

add_num(5)

15

#### Writing a Factorial Function

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [32]:
def subtract_num(num):
    if num == 1:
        print(f'subtract_num({num}) = 1')
        return num
    print(f'subtract_num({num}) = {num} - subtract_num({num - 1})')
    return num - subtract_num(num - 1)

subtract_num(10)

subtract_num(10) = 10 - subtract_num(9)
subtract_num(9) = 9 - subtract_num(8)
subtract_num(8) = 8 - subtract_num(7)
subtract_num(7) = 7 - subtract_num(6)
subtract_num(6) = 6 - subtract_num(5)
subtract_num(5) = 5 - subtract_num(4)
subtract_num(4) = 4 - subtract_num(3)
subtract_num(3) = 3 - subtract_num(2)
subtract_num(2) = 2 - subtract_num(1)
subtract_num(1) = 1


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
# use Filter
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]



### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
# use Lambda
# key argument on .sort() method

author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [ ]:
# use Map
# F = (9/5)*C + 32

places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [ ]:
# use Recursion

